docs/syscalls.md

```diff
@@ -0,0 +1,241 @@
+The Demikernel System Call Interface
+=====================================
+
+This document provides a detailed description of the Demikernel system call
+interface.
+
+Overall, all system calls return zero if successful and an error code otherwise.
+Furthermore, each system call may take incoming and/or outgoing arguments. In
+the description listed next, incoming arguments are labelled `(in)`, whereas
+outgoing arguments are labelled `(out)`.
+
+Finally, some system calls may be tagged to indicate some point of attention:
+- `deprecated`: the system call shall be dropped in future releases of Demikernel
+- `missing`: the underlying implementation of that system call is not yet exposed
+
+
+System Calls in `include/dmtr/libos.h`
+--------------------------------------
+
+This file contains most of the system calls that operate on the control and data
+paths.
+
+----
+
+* `int dmtr_init(int argc, char *argv[]);`
+  * `argc` (in) : number of command line arguments, passed on to the libOS
+  * `argv` (in) : values of command line arguments, passed on to the libOS
+
+Initializes libOS state. Sets up devices, allocates data structures and performs
+general initialization tasks.
+
+----
+
+* `int dmtr_queue(int *qd_out);`
+  * `qd_out`(out) : queue descriptor for newly allocated memory queue
+    if successful; otherwise invalid
+
+ `(missing)` Allocates an in-memory Demikernel queue. Returns a queue descriptor
+ that references a FIFO queue (i.e., popping an element from the queue will
+ return the first scatter-gather array that was pushed).
+
+----
+
+* `int dmtr_socket(int *qd_out, int domain, int type, int protocol);`
+  * `qd_out` (out) : queue descriptor for newly allocated network
+    queue connected to the socket if successful; otherwise invalid
+  * `domain` (in) : communication domain for the newly allocated
+    socket queue if appropriate for the libOS
+  * `type` (in) : type for the newly allocated socket queue if
+    appropriate; sometimes translated by the libOS. For example, Linux
+    RDMA will allocate an RC queue pair if the type is `SOCK_STREAM`
```

Comment:
Doesn't need to specify Linux RDMA. Technically our Windows RDMA libOS does the same, although it's not publicly available/supported